### DISCUSSION TITLE
Domande riguardo alla suddivisione dei vari utenti

### DISCUSSION THREAD

**Discussion by Giann8** [sidelux/LootBot#8] [Q&A]
Ciao a tutti, sono nuovo nell'ambito node.js. Sto cercando di creare un bot simile a lootbot custom da utilizzare con degli amici e per programmare qualcosina, nel farlo sto utilizzando telegraf.js e MongoDB. Non ho problemi con il database e con telegraf nemmeno, ho un problema però a suddividere le istanze degli utenti. Quando il bot viene startato in automatico prende lo username del giocatore in modo da copiare dal database il profilo (inventario, esperienza ecc). Unico problema è che se si ha più di un giocatore il bot va ad utilizzare sempre lo username dell'ultima persona che ha startato quindi se il primo avvia una missione o fa altro tutto viene considerato svolto dal secondo. Ho provato ad utilizzare  una mappa ma il problema non cambia.
Vi allego parte del codice,  la mappa è una variabile globale.
Per domande riguardo al codice scrivetemi pure, ho assolutamente bisogno di un aiuto.
```
bot.start(async (ctx) => {
    var User = new Character
    User.usern = ctx.message.from.username

    let GetId = () => { return ctx.chat.id }

    if (User.usern == null) {
        return ctx.sendMessage('A quanto pare non hai impostato alcuno username, perfavore vai nelle impostazioni di telegram e crea un nuovo username poi torna e utilizza /start')
    }

    let Player = await dbo.findOne({ usern: User.usern })
    if (Player == null) {
        dbo.updateOne(
            { usern: User.usern },
            { $set: { usern: User.usern, name: ctx.message.from.first_name, level: User.level, exp: User.exp, backpack: User.backpack } },
            { upsert: true }
        )
        return ctx.sendMessage('Hey benvenuto, a quanto pare sei nuovo restarta il bot prima di continuare con /start')
    }
    User = Player

    InGameUsers.set(GetId, User)

    ctx.scene.enter('Università_id')

```

**Comment by sidelux** [comment]
Ciao, mi servirebbe vedere come funziona InGameUsers (la funzione set), in generale così senza approfondire credo che lo User che vai ad usare venga sovrascritto ogni volta che il bot viene startato e quindi viene usato sempre l'ultimo, dovresti usare un approccio tipo User[ctx.message.from.id] al posto dell'User e basta.

**Reply by Giann8** [reply]
la funzione set va ad appendere in fondo a InGameUsers lo User(quindi l'oggetto player) avendo come chiave l'id della chat dello user stesso in pratica facendo InGameUsers.set(GetId1,User1) allora InGameUsers.get(GetId1)=User1. la posizione degli user non cambia, per trovarli serve usare solo la chiave. Questo è ciò che viene restituito da InGameUsers quando una sola persona starta il bot:

```
//Console.log('post caricamento',InGameUsers) 
post caricamento Map(1) {
  [Function: GetId] => {
    _id: new ObjectId("647f08f63d01a6868d774533"),
    usern: 'The_Gamer_morty',
    backpack: { Item: [Array], Numberitems: 3 },
    exp: 10,
    level: 2,
    name: 'Gra',
    'facoltà': 'Informatica 💻'
  }
}
//Console.log(InGameUsers.get(getId)) 
{
  _id: new ObjectId("647f08f63d01a6868d774533"),
  usern: 'The_Gamer_morty',
  backpack: { Item: [ [Object] ], Numberitems: 3 },
  exp: 10,
  level: 2,
  name: 'Gra',
  'facoltà': 'Informatica 💻'
}
```
Quando anche la seconda stata il bot viene restituito di consueto il nome del secondo:
 ```
[Function: GetId] => {
    _id: new ObjectId("647f08f63d01a6868d774533"),
    usern: 'The_Gamer_morty',
    backpack: { Item: [Array], Numberitems: 5 },
    exp: 10,
    level: 4,
    name: 'Gra',
    'facoltà': 'Informatica 💻'
  },
  [Function: GetId] => {
    _id: new ObjectId("647df2c73d01a6868ddb1f75"),
    usern: 'Passo',
    backpack: { Item: [Array], Numberitems: 1 },
    exp: 30,
    level: 2,
    name: 'Passo',
    'facoltà': 'Informatica 💻'
  }
}
////Console.log(InGameUsers.get(getId)) 

{
  _id: new ObjectId("647df2c73d01a6868ddb1f75"),
  usern: 'Passo',
  backpack: { Item: [ [Object] ], Numberitems: 1 },
  exp: 30,
  level: 2,
  name: 'Passo',
  'facoltà': 'Informatica 💻'
}

```
mentre dalla mappa viene preso lo user giusto nella scena entra la persona sbagliata:
```
 Università.enter(async (ctx) => {


        await ctx.replyWithHTML(`Ciao <b>${InGameUsers.get(GetId).usern}</b>.\n🎴Livello: ${InGameUsers.get(GetId).level}\n📚esperienza: ${InGameUsers.get(GetId).exp} \n❖Eventi: <b>${EventoOn}</b>\nMissione: <b>${CheckMission(InGameUsers.get(GetId))}</b>\nCosa vuoi fare oggi? `, Markup.keyboard(Uni).resize(true))
}
```
Cioè questa linea di codice viene inviata allo start alla persona giusta ma presenta lo user sbagliato, invece di restituire "Ciao Passo..." restituisce "ciao The_gamer_morty...", con anche il resto delle statistiche sbagliate. Questo va ad influire anche poi nelle missioni.

**Reply by sidelux** [reply]
L'unica cosa che mi viene in mente allora è che la funzione di get vada a tirare fuori l'utente sbagliato, se vuoi prova ad inviare anche quello così guardo.

**Reply by Giann8** [reply]
in realtà la funzione .get() è interna a javascript, potrebbe essere che sia getId quella sbagliata? mi sembra strano perché funziona comunque nello start

**Reply by sidelux** [reply]
Potrebbe sì, ti consiglio di isolare quella funzione assicurandoti che tutte le altre funzioni loggandole funzionino correttamente, se getId come immagino ritorna un utente errato allora l'errore è al suo interno.

**Reply by Giann8** [reply]
allora controllando con un amico abbiamo fatto qualche prova, adesso il messaggio del menù con "ciao user..." viene sempre giusto, il problema rimane per le missioni e per lo zaino, ho tolto il getId() direttamente rimane comunque il problema in queste due scene, ti lascio il codice per le missioni(lessonId fa la stessa cosa di ctx ma ha nome diverso):
```
 Lezione.enter((lessonId) => {
            lessonId.reply('Vuoi iniziare una nuova lezione?', Markup.keyboard(comm).resize(true))
            bot.hears('Si', async () => {
                console.log(InGameUsers.get((lessonId.chat.id)).mission)
                if (InGameUsers.get(lessonId.chat.id).mission === undefined) {
                    InGameUsers.get(lessonId.chat.id).mission = false
                }
                if (InGameUsers.get(lessonId.chat.id).mission === false) {
                    InGameUsers.get(lessonId.chat.id).mission = true
                    var oggetto = await StartMission(ms, InGameUsers.get(lessonId.chat.id), oggetto)
                    setTimeout(function () { ctx.replyWithHTML('Missione terminata. Hai ottenuto <b>' + oggetto.name + '</b>!') }, ms)
                    setTimeout(function () { return InGameUsers.get(lessonId.chat.id).mission = false }, ms)
                    lessonId.sendMessage('Missione iniziata. Terminerà alle ' + min)
                } else {
                    lessonId.sendMessage('Pazzo, sei già a lezione')
                }
                lessonId.scene.enter('Università_id')
            })
```
ho messo un console.log ogni volta che qualcuno torna nella scena 'università' a quanto pare io possiedo spesso il player dell'altro user
ah e grazie mille dell'aiuto, non vorrei rubarti troppo tempo.

**Reply by sidelux** [reply]
In questo caso credo che il problema stia nel fatto che il bot.hears venga istanziato solo una volta, non ogni volta che si entra nella funzione, quindi il lessonId è sempre uguale.
Per risolvere questo problema io ho utilizzato una tecnica un po' macchinosa ma che ha funzionato, la trovi in molti punti del mio codice cercando ad esempio questa riga:
`answerCallbacks[message.chat.id] = async function (answer) {

if (answer.text.toLowerCase() == "si") {`

**Reply by Giann8** [reply]
una cosa, ho cercato la riga di codice che mi hai detto e ho notato che usi spesso connection.query, vai quindi ad estrarre ogni volta dalla collection le statistiche del player? Chiedo perché in mongodb non esiste il comando connection.query

**Reply by sidelux** [reply]
Sì esatto, perchè il mio bot deve sopravvivere anche quando viene riavviato, altrimenti ogni volta perderei le utenze e i dati.

**Reply by Giann8** [reply]
Allora armeggiando un po' ho scoperto che per ctx esistono due modi per ottenere l'id della chat, o con ctx.chat.id o con ctx.message.from.id, io utilizzavo la prima che però prendeva l'id della chat in cui era andando a mischiare il tutto, utilizzando message.from però funziona tutto. Pensavo fosse un problema più complesso invece era una cavolata. grazie mille per il tuo tempo.

**Reply by sidelux** [reply]
Ah non avendo mai usato Telegraf non sapevo, bene così!

**Reply by Giann8** [reply]
Nemmeno io, avevo provato a cercare nella documentazione ma è fatta troppo male, anche online nessuno fa riferimento a questo.

**Reply by nrc382** [reply]
Mi intrometto:

Più che la documentazione di telegraf ti consiglio di vedere quella delle api di telegram (ed in particolare come è costruito l'oggetto message)

https://core.telegram.org/bots/api#message

**Reply by Giann8** [reply]
ah okok grazie mille

**Comment by nrc382** [comment]
> Ciao a tutti, sono nuovo nell'ambito node.js. Sto cercando di creare un bot simile a lootbot custom da utilizzare con degli amici e per programmare qualcosina, nel farlo sto utilizzando telegraf.js e MongoDB. Non ho problemi con il database e con telegraf nemmeno, ho un problema però a suddividere le istanze degli utenti. Quando il bot viene startato in automatico prende lo username del giocatore in modo da copiare dal database il profilo (inventario, esperienza ecc). Unico problema è che se si ha più di un giocatore il bot va ad utilizzare sempre lo username dell'ultima persona che ha startato quindi se il primo avvia una missione o fa altro tutto viene considerato svolto dal secondo. Ho provato ad utilizzare  una mappa ma il problema non cambia.
> 
> Vi allego parte del codice,  la mappa è una variabile globale.
> 
> Per domande riguardo al codice scrivetemi pure, ho assolutamente bisogno di un aiuto.
> 
> ```
> 
> bot.start(async (ctx) => {
> 
>     var User = new Character
> 
>     User.usern = ctx.message.from.username
> 
> 
> 
>     let GetId = () => { return ctx.chat.id }
> 
> 
> 
>     if (User.usern == null) {
> 
>         return ctx.sendMessage('A quanto pare non hai impostato alcuno username, perfavore vai nelle impostazioni di telegram e crea un nuovo username poi torna e utilizza /start')
> 
>     }
> 
> 
> 
>     let Player = await dbo.findOne({ usern: User.usern })
> 
>     if (Player == null) {
> 
>         dbo.updateOne(
> 
>             { usern: User.usern },
> 
>             { $set: { usern: User.usern, name: ctx.message.from.first_name, level: User.level, exp: User.exp, backpack: User.backpack } },
> 
>             { upsert: true }
> 
>         )
> 
>         return ctx.sendMessage('Hey benvenuto, a quanto pare sei nuovo restarta il bot prima di continuare con /start')
> 
>     }
> 
>     User = Player
> 
> 
> 
>     InGameUsers.set(GetId, User)
> 
> 
> 
>     ctx.scene.enter('Università_id')
> 
> 
> 
> ```

Mi permetto anche di suggerirti di non usare questo approccio scrivendo il tuo bot. 
In pochissimo tempo potresti non essere più in grado di leggere il tuo stesso codice. 

Prova a dare un occhiata a come ho impostato questo bot (non uso telegraf, ma il concetto resta identico)

https://github.com/nrc382/Bardibot/tree/main/Bardi

**Reply by Giann8** [reply]
con questo approccio intendi che dovrei dividere su più file il programma?

**Reply by nrc382** [reply]
> con questo approccio intendi che dovrei dividere su più file il programma?

in generale dividere logica, presentazione dati e persistenza ti aiuta a tenere il codice scalabile (ma anche a suggerirti come approcciarti ai problemi che ti si pongono).
Nel particolare di bot telegram io penso sia una buona idea quella di separare ulteriormente anche le stringhe usate per riempire i messaggi. 

Javascript può prestarsi quasi a qualsiasi framework. 
Se non hai grandi basi ti suggerisco di vedere qualche video su programmazione atta a messaggi e oggetti (oop) e quella funzionale (fp) e come integrarle in un approccio mvc :)

Ti scrivo queste cose perché penso possano davvero aiutare a programmare in maniera più efficace e soprattutto divertente
(Potendo fare praticamente ogni cosa in ogni modo immaginabile e non è facilissimo perdersi dopo poco, e questo spesso è frustrante.)


Ah!
E **commenta** il tuo codice!
Io faccio ancora fatica ad abituarmi a questa cosa, ma è davvero importante (per il te di fra qualche tempo)

**Reply by Giann8** [reply]
si in effetti dovrei proprio guardare qualche video, diciamo che ho usato i commenti per dividere i vari pezzi in app.js. scherzi a parte effettivamente commentare lo faccio relativamente poco anche per esercizi universitari, mentre la suddivisione  del programma in generale l’ho fatta solo per le classi e per la funzione startmission, mi spaventava soprattutto rischiare di rompere il programma creando un file a parte per l’update magari dei dati nelle collection del database o per le scene del bot. Cercherò qualche video e mi metterò in pari, grazie ancora per i consigli e l’aiuto

**Reply by nrc382** [reply]
> si in effetti dovrei proprio guardare qualche video, diciamo che ho usato i commenti per dividere i vari pezzi in app.js. scherzi a parte effettivamente commentare lo faccio relativamente poco anche per esercizi universitari, mentre la suddivisione  del programma in generale l’ho fatta solo per le classi e per la funzione startmission, mi spaventava soprattutto rischiare di rompere il programma creando un file a parte per l’update magari dei dati nelle collection del database o per le scene del bot. Cercherò qualche video e mi metterò in pari, grazie ancora per i consigli e l’aiuto

Per l'università è una pena...
Ma se ci si forza ad abituarsi prima di scrivere codice a descrivere in linguaggio naturale quello che si vuole... beh i vantaggi sono tantissimi. 

Ps
Sei in "playground". Non avere paura di rompere cose, a questo stage ne avresti solo da imparare :)

Auguri!
